### PR TITLE
IDENTITY-5424: Ability to Configure Regex Patterns for Callback URLs when working with OIDC Session Management 

### PR DESCRIPTION
**Pull request by harsha1979**
### Ability to Configure Regex Patterns for Callback URLs when working with OIDC Session Management

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServlet.java [hunk 1]
```diff
@@ -37,31 +57,80 @@ public void init() throws ServletException {
...
+            String callbackURL = getCallbackURL(request, clientId);
+            String clientOrigin = OIDCSessionManagementUtil.getOrigin(callbackURL);
+
             if (log.isDebugEnabled()) {
-                log.debug("Invalid request.\'client_id\' not found in request as parameter");
+                log.debug("Client Origin : " + clientOrigin);
             }
+            response.getWriter().print(getOPIFrame(clientOrigin));
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            log.error("Error while retrieving OAuth application information for the provided client id, "+  e.getMessage
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServlet.java]: logging the client_id here would be better

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServlet.java [hunk 2]
```diff
@@ -37,31 +57,80 @@ public void init() throws ServletException {
...
 
         response.setContentType("text/html");
 
         String clientId = request.getParameter(OIDCSessionConstants.OIDC_CLIENT_ID_PARAM);
-        if (StringUtils.isBlank(clientId)) {
+
+        try {
+            if (StringUtils.isBlank(clientId)) {
+                throw new OIDCSessionManagerException(
+                        "Invalid request. client_id not found in request as parameter.");
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServlet.java]: don't we have any context information like tenant domain

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServlet.java [hunk 3]
```diff
@@ -37,31 +57,80 @@ public void init() throws ServletException {
...
+        if (log.isDebugEnabled()) {
+            log.debug("Requested client_id : " + clientId + " Configured callbackUrl : " + configuredCallbackURL);
+        }
+        if (StringUtils.isBlank(configuredCallbackURL)) {
+            throw new OIDCSessionManagerException(
+                    "CallbackURL is empty in service provider configuration, clientId : " + clientId);
+        }
+        if (configuredCallbackURL.startsWith(OAuthConstants.CALLBACK_URL_REGEXP_PREFIX)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Regex value found for callback url in service provider.");
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServlet.java]: add context information for debug log

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServlet.java [hunk 4]
```diff
@@ -37,31 +57,80 @@ public void init() throws ServletException {
...
+            } else {
                 if (log.isDebugEnabled()) {
-                    log.debug("Error while retrieving OAuth application information for the provided client id", e);
+                    log.debug("Requested redirect_uri from rp IFrame : " + rpIFrameReqCallbackURL);
+                }
+                String regexp = configuredCallbackURL
+                        .substring(OAuthConstants.CALLBACK_URL_REGEXP_PREFIX.length());
+                if (rpIFrameReqCallbackURL.matches(regexp)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Requested redirect_uri is matched with the regex in service provider.");
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCSessionIFrameServlet.java]: add context information

**Pull request merged**